src/coreclr/nativeaot/Runtime/threadstore.cpp

```diff
@@ -248,7 +249,13 @@ void ThreadStore::SuspendAllThreads(bool waitForGCEvent)
     	        // @TODO: need tuning for spin
                 // @TODO: need tuning for this whole loop as well.
                 //        we are likley too aggressive with interruptions which may result in longer pauses.
-                YieldProcessorNormalizedForPreSkylakeCount(normalizationInfo, 10000);
+                YieldProcessorNormalizedForPreSkylakeCount(normalizationInfo, waitCycles);
+
+                // simplistic linear backoff for now
+                // we could be catching threads in restartable sequences such as LL/SC style interlocked on ARM64
+                // and forcing them to restart.
+                // if interrupt mechanism is fast, eagerness could be hurting our overall progress.
+                waitCycles += 10000;
```

Comment:
Ultimately the wait is unbounded, since suspension cannot gracefully fail. - "It is done, when it is done". 
Practically, we can have very good timings here.

Both waiting too long and waiting not enough between re-hijacking can make the whole thing last longer.
Here I observed that we would interrupt a thread doing LL/SC loop, which would invalidate its monitor, so it would fail and try again - just in time for us to interrup it again - leading to very long hangs.
Fixed spin counts are often a bad idea as the guesstimate may be wrong when running on a different platform. I just made the spin count to adjust in a naive way for now.

Dealing with this loop is the next/last part in the NativeAOT suspension work item. https://github.com/dotnet/runtime/issues/67805

I plan to make this similar to what CoreCLR does:
- interrupt threads and hijack/suspend accordingly 
- spin-wait without re-hijacking as we are making progress - the common case is everything suspends quickly.
- otherwise wait for progress with a 1msec. timeout, and then try again - to deal with remaining strugglers.

Reply:
Makes sense